internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,304 @@
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
+	numControllers int,
+	numLUNsPerController int,
+	guestMountFmt string,
+	reservedSlots []Slot,
+) *Manager {
+	am := newAttachManager(attacher, unplugger, numControllers, numLUNsPerController, reservedSlots)
+	mm := newMountManager(mounter, guestMountFmt)
+	return &Manager{am, mm}
+}
+
+// MountConfig specifies the options to apply for mounting a SCSI device in
+// the guest OS.
+type MountConfig struct {
+	Partition uint64
+	Encrypted bool
+	Options   []string
+}
+
+// Mount represents a SCSI device that has been attached to a VM, and potentially
+// also mounted into the guest OS.
+type Mount struct {
+	mgr         *Manager
+	controller  uint
+	lun         uint
+	guestPath   string
+	releaseOnce sync.Once
+}
+
+// Controller returns the controller number that the SCSI device is attached to.
+func (m *Mount) Controller() uint {
+	return m.controller
+}
+
+// LUN returns the LUN number that the SCSI device is attached to.
+func (m *Mount) LUN() uint {
+	return m.lun
+}
+
+// GuestPath returns the path inside the guest OS where the SCSI device was mounted.
+// Will return an empty string if no guest mount was performed.
+func (m *Mount) GuestPath() string {
+	return m.guestPath
+}
+
+// Release releases the SCSI mount. Refcount tracking is used in case multiple instances
+// of the same attachment or mount are used. If the refcount for the guest OS mount
+// reaches 0, the guest OS mount is removed. If the refcount for the SCSI attachment
+// reaches 0, the SCSI attachment is removed.
+func (m *Mount) Release(ctx context.Context) (err error) {
+	err = ErrAlreadyReleased
+	m.releaseOnce.Do(func() {
+		err = m.mgr.remove(ctx, m.controller, m.lun, m.guestPath)
+	})
+	return
+}
+
+// AddVirtualDisk attaches and mounts a VHD on the host to the VM. If the same
+// VHD has already been attached to the VM, the existing attachment will
+// be reused. If the same VHD has already been mounted in the guest OS
+// with the same MountConfig, the same mount will be reused.
+//
+// If vmID is non-empty an ACL will be added to the VHD so that the specified VHD
+// can access it.
+//
+// mc determines the settings to apply on the guest OS mount. If
+// it is nil, no guest OS mount is performed.
+func (m *Manager) AddVirtualDisk(
+	ctx context.Context,
+	hostPath string,
+	readOnly bool,
+	vmID string,
+	mc *MountConfig,
+) (*Mount, error) {
+	if m == nil {
+		return nil, ErrNotInitialized
+	}
+	if vmID != "" {
+		if err := wclayer.GrantVmAccess(ctx, vmID, hostPath); err != nil {
+			return nil, err
+		}
+	}
+	var mcInternal *mountConfig
+	if mc != nil {
+		mcInternal = &mountConfig{
+			readOnly:  readOnly,
+			encrypted: mc.Encrypted,
+			options:   mc.Options,
+			verity:    readVerityInfo(ctx, hostPath),
```

Comment:
These are all missing setting the partition from `MountConfig`

Reply:
Ah, this got lost in some rebasing I think.

Reply:
Fixed.